juliet/src/protocol/outgoing_message.rs

```diff
@@ -0,0 +1,431 @@
+
+use super::payload_is_multi_frame;
+
+/// A message to be sent to the peer.
+///
+/// [`OutgoingMessage`]s are generated when the protocol requires data to be sent to the peer.
+/// Unless the connection is terminated, they should not be dropped, but can be sent in any order.
+///
+/// While *frames* can be sent in any order, a message may span one or more frames, which can be
+/// interspersed with other messages at will. In general, the [`OutgoingMessage::frames()`] iterator
+/// should be used, even for single-frame messages.
+#[must_use]
+#[derive(Debug)]
+pub struct OutgoingMessage {
+    /// The common header for all outgoing messages.
+    header: Header,
+    /// The payload, potentially split across multiple messages.
+    payload: Option<Bytes>,
+}
+
+impl OutgoingMessage {
+    /// Constructs a new outgoing messages.
+    // Note: Do not make this function available to users of the library, to avoid them constructing
+    //       messages by accident that may violate the protocol.
+    #[inline(always)]
+    pub(super) fn new(header: Header, payload: Option<Bytes>) -> Self {
+        Self { header, payload }
+    }
+
+    /// Returns whether or not a message will span multiple frames.
+    #[inline(always)]
+    pub fn is_multi_frame(&self, max_frame_size: u32) -> bool {
+        if let Some(ref payload) = self.payload {
+            payload_is_multi_frame(max_frame_size, payload.len())
+        } else {
+            false
+        }
+    }
+
+    /// Creates an iterator over all frames in the message.
+    #[inline(always)]
+    pub fn frames(self) -> FrameIter {
+        FrameIter {
+            msg: self,
+            bytes_processed: 0,
+        }
+    }
+
+    /// Returns the outgoing message's header.
+    #[inline(always)]
+    pub fn header(&self) -> Header {
+        self.header
+    }
+}
+
+/// Combination of header and potential frame payload length.
+///
+/// A message with a payload always start with an initial frame that has a header and a varint
+/// encoded payload length. This type combines the two, and allows for the payload length to
+/// effectively be omitted (through [`Varint32::SENTINEL`]). It has a compact, constant size memory
+/// representation regardless of whether a variably sized integer is present or not.
+///
+/// This type implements [`AsRef<u8>`], which will return the correctly encoded bytes suitable for
+/// sending header and potential varint encoded length.
+#[derive(Clone, Copy, Debug, Pod, Zeroable)]
+#[repr(C)]
+struct Preamble {
+    /// The header, which is always sent.
+    header: Header,
+    /// The payload length. If [`Varint32::SENTINEL`], it will always be omitted from output.
+    payload_length: Varint32,
+}
+
+impl Display for Preamble {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Display::fmt(&self.header, f)?;
+        if !self.payload_length.is_sentinel() {
+            write!(f, " [l={}]", self.payload_length.decode())?;
```

Comment:
I think this is too opaque.  Could we stretch to two more chars?
```suggestion
            write!(f, " [len={}]", self.payload_length.decode())?;
```

Reply:
What, and double computational resources spent from 2 bytes to 4 bytes??

Reply:
Done.